### PR TITLE
Enhance VideoInputOption on Windows: …

### PR DESCRIPTION
**Pull request by ChristopheI**
Get VCodec (h264 and mjpeg) or InputFormat (pixel format in fact)

Fix also how FrameRate is set

Related to issue #33

### CODE REVIEW

File: SeeShark/Utils/DShowUtils.cs [hunk 1]
```diff
@@ -71,15 +71,14 @@ private unsafe static List<VideoInputOptions> getAvailableOptions(DsDevice dsDev
...
-                while (fetched != IntPtr.Zero && mediaTypes[0] != null)
+                while (mediaTypes[0] != null)
                 {
                     Marshal.PtrToStructure(mediaTypes[0].formatPtr, v);
 
                     if (v.BmiHeader.Size != 0 && v.BmiHeader.BitCount != 0)
                     {
                         if (v.BmiHeader.BitCount > bitCount)
                         {
-                            options.Clear();
```
Comment (Speykious) [inline SeeShark/Utils/DShowUtils.cs:82]: About this line, I'm not even sure why it was here in the code examples I found, but I'm even less sure of why it shouldn't be here.
Given your current code that seems to work, can you try putting this line again and see what happens?

File: SeeShark/Utils/DShowUtils.cs [hunk 2]
```diff
@@ -105,23 +104,52 @@ private unsafe static List<VideoInputOptions> getAvailableOptions(DsDevice dsDev
...
 
-                        options.Add(new VideoInputOptions
+                        var vio = new VideoInputOptions
                         {
-                            InputFormat = new string((sbyte*)codec->name),
                             VideoSize = (v.BmiHeader.Width, v.BmiHeader.Height),
-                            Framerate = new AVRational
+                            // https://docs.microsoft.com/en-us/windows/win32/directshow/configure-the-video-output-format
+                            // "frames per second = 10,000,000 / frame duration"
+                            Framerate = ffmpeg.av_d2q((double)10_000_000L / v.AvgTimePerFrame, (int)10_000_000L),
```
Comment (Speykious) [inline SeeShark/Utils/DShowUtils.cs:113]: Good function, I didn't know about it :+1:

File: SeeShark/Utils/DShowUtils.cs [hunk 3]
```diff
@@ -105,23 +104,52 @@ private unsafe static List<VideoInputOptions> getAvailableOptions(DsDevice dsDev
...
+                                    vio.VCodec = "h264";
+                                }
+                                else if (Guid.Equals(mediaTypes[0].subType, MediaSubType.MJPG))
+                                {
+                                    vio.VCodec = "mjpeg";
+                                }
+                                else
+                                {
+                                    Console.WriteLine(mediaTypes[0].subType);
+                                }
```
Comment (Speykious) [inline SeeShark/Utils/DShowUtils.cs:143]: I'll probably put that code in a separate method. I'll see what I can do

File: SeeShark/VideoInputOptions.cs [hunk 4]
```diff
@@ -39,9 +39,15 @@ public class VideoInputOptions
...
-        /// If the video stream is raw, it is the name of its pixel format, otherwise it is the name of its codec.
+        /// On all platform (except Windows), if the video stream is raw, it is the name of its pixel format, otherwise it is the name of its codec.
+        /// On Windows, contains only the pixel format. (can be null/empty)
         /// </summary>
         public string? InputFormat { get; set; }
+        /// <summary>
+        /// Used on Windows only, contains the codec (can be null/empty)
+        /// If VCodec is not null/empty, InputFormat is null/empty (and the opposite)
+        /// </summary>
+        public string? VCodec { get; set; }
```
Comment (Speykious) [inline SeeShark/VideoInputOptions.cs:50]: It seems that you always specify _either_ the `vcodec` or the `pixel_format`. We can probably avoid putting a separate field for the `vcodec` in this case, I'd rather have SeeShark deal with that automatically in code to not complicate the API.

**Review by Speykious — CHANGES_REQUESTED** [review]

File: SeeShark/Utils/DShowUtils.cs (hunk 1 above)
Comment (ChristopheI) [reply SeeShark/Utils/DShowUtils.cs:82]: When used I have this output for my second webcam: 
image

So less results 16 instead 52: see here https://github.com/vignetteapp/SeeShark/issues/33#issuecomment-1259635656

File: SeeShark/VideoInputOptions.cs (hunk 4 above)
Comment (ChristopheI) [reply SeeShark/VideoInputOptions.cs:50]: Correct. 
Perhaps **pixel_format** could be used like this (on windows)
- pixel_format => yuyv422, rgb24 , ... (so the pixel format name)
- pixel_format => vcodec=mjpeg, vcodec=h264 (so the video format with "vcodec" as prefix

**Review by Speykious — APPROVED** [review]
It's time we finish #30

**Pull request merged**